### DISCUSSION TITLE
Python CDK Setup

### DISCUSSION THREAD

**Discussion by dbanty** [dbanty/dylananthony.com#36] [Ideas]
I've recently put a bunch of work into making my APIGateway + Lambda setup in CDK work with Python applications. The hardest part was getting the dependencies all building correctly. If you're interested in a blog post / tutorial on this topic, upvote this thread! If there are specific things you'd like to see or questions you'd like answered on this topic, comment them below.

**Comment by mosheduminer** [comment]
It happens that right now I'm interested in learning about the CDK in general. Would definitely appreciate a blog post!

BTW I'm curious about typing in the python CDK. Does python's relatively immature type system result in bugs when using the CDK? What are the results?

**Reply by dbanty** [reply]
Yeah there are some rough patches for sure. CDK does include type annotations, but I think they’re all generated from the equivalent TypeScript code. 

In TS, they use interfaces for props so you don’t need direct inheritance. In Python, rather than using `Protocol`—which is the closest equivalent to interfaces—they use a normal class. My guess is that’s to support older Python versions.

Unfortunately, even without using `Protocol`, they still want the flexibility of interfaces. So they don’t inherit from the not-interface in their concrete implementations. All of this is to say it looks like you’re passing in the wrong type to a lot of classes/functions, so you really can’t rely on the static types at all. Instead you get IDE/mypy errors while writing 🙁.

This was annoying enough that I actually stuck with TypeScript for writing the CDK code of my Python lambdas at work.

**Reply by mosheduminer** [reply]
> All of this is to say it looks like you’re passing in the wrong type to a lot of classes/functions, so you really can’t rely on the static types at all.

> This was annoying enough that I actually stuck with TypeScript for writing the CDK code of my Python lambdas at work.

This was my worst fear. When I start actually writing CDK code, I'll probably stick to TS as well.

**Reply by dbanty** [reply]
The cdk cli tool requires node anyway, so using TS doesn’t feel _too_ bad. I just include the CLI in that same package.json.
